python/ray/data/impl/shuffle.py

```diff
@@ -13,137 +13,161 @@
-    input_num_blocks = len(input_blocks)
-
-    shuffle_map = cached_remote_fn(_shuffle_map)
-    shuffle_reduce = cached_remote_fn(_shuffle_reduce)
-
-    map_bar = ProgressBar("Shuffle Map", position=0, total=input_num_blocks)
-
-    shuffle_map_out = [
-        shuffle_map.options(
-            **map_ray_remote_args,
-            num_returns=1 + output_num_blocks,
-        ).remote(block, block_udf, i, output_num_blocks, random_shuffle, random_seed)
-        for i, block in enumerate(input_blocks)
-    ]
-
-    # The first item returned is the BlockMetadata.
-    shuffle_map_metadata = []
-    for i, refs in enumerate(shuffle_map_out):
-        shuffle_map_metadata.append(refs[0])
-        shuffle_map_out[i] = refs[1:]
-
-    # Eagerly delete the input block references in order to eagerly release
-    # the blocks' memory.
-    del input_blocks
-    shuffle_map_metadata = map_bar.fetch_until_complete(shuffle_map_metadata)
-    map_bar.close()
-
-    # Randomize the reduce order of the blocks.
-    if random_shuffle:
-        random = np.random.RandomState(random_seed)
-        random.shuffle(shuffle_map_out)
-
-    reduce_bar = ProgressBar("Shuffle Reduce", position=0, total=output_num_blocks)
-    shuffle_reduce_out = [
-        shuffle_reduce.options(
-            **reduce_ray_remote_args,
-            num_returns=2,
-        ).remote(*[shuffle_map_out[i][j] for i in range(input_num_blocks)])
-        for j in range(output_num_blocks)
-    ]
-    # Eagerly delete the map block references in order to eagerly release
-    # the blocks' memory.
-    del shuffle_map_out
-    new_blocks, new_metadata = zip(*shuffle_reduce_out)
-    reduce_bar.block_until_complete(list(new_blocks))
-    new_metadata = ray.get(list(new_metadata))
-    reduce_bar.close()
-
-    stats = {
-        "map": shuffle_map_metadata,
-        "reduce": new_metadata,
-    }
-
-    return BlockList(list(new_blocks), list(new_metadata)), stats
-
-
-def _shuffle_map(
-    block: Block,
-    block_udf: Optional[Callable[[Block], Iterable[Block]]],
-    idx: int,
-    output_num_blocks: int,
-    random_shuffle: bool,
-    random_seed: Optional[int],
-) -> List[Union[BlockMetadata, Block]]:
-    """Returns list of [BlockMetadata, O1, O2, O3, ...output_num_blocks]."""
-    stats = BlockExecStats.builder()
-    if block_udf:
-        # TODO(ekl) note that this effectively disables block splitting.
-        blocks = list(block_udf(block))
-        if len(blocks) > 1:
-            builder = BlockAccessor.for_block(blocks[0]).builder()
-            for b in blocks:
-                builder.add_block(b)
-            block = builder.build()
-        else:
-            block = blocks[0]
-    block = BlockAccessor.for_block(block)
-
-    # Randomize the distribution of records to blocks.
-    if random_shuffle:
-        seed_i = random_seed + idx if random_seed is not None else None
-        block = block.random_shuffle(seed_i)
+class ShuffleOp:
```

Comment:
One alternative that doesn't need to define classes is that we just define a function:

```
shuffle_map_reduce(input_blocks, output_num_blocks, clear_input_blocks, map_fn, map_remote_args, reduce_fn, reduce_remote_args) {
    // current execute implementation
}
```
We can also accept map_args and reduce_args explicitly or they can be captured in the map_fn and reduce_fn.

I feel that might be cleaner.

Reply:
I think I prefer the class-based version for this to make the map_fn and reduce_fn signature requirements clear.

Reply:
I personally prefer the class-based version as well for that reason, but I'm not too bullish on that.